### ISSUE TITLE
About how to put data with Batch dimension into GAT？

### ISSUE THREAD

**Issue by WangYibucea** [rusty1s/pytorch_geometric#3044]
Hello, i meet some question as follow:
Currently, 
my data dimension is (batch, nodes, feature), 
and the adjacency matrix dimension is (batch, nodes, nodes)
At this point, both the data and the adjacency matrix do not meet the requirements of GAT.
It is possible to do a loop on Batch into GAT, but the operation will be very slow.
Is there any way to put the data with Batch dimension into GAT directly without looping Batch?

**Comment by rusty1s** [comment]
PyG uses a different format for learning in mini-batches, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#mini-batches), in which node features are concatenated along the node dimension, and adjacency matrices are stacked diagonally. This is automatically supported when using the `torch_geometric.data.DataLoader` class.

**Comment by WangYibucea** [comment]
Hello, thank you very much for your answer. I am currently using HyperGCN. since the graph structure is generated during the training process, it is not convenient to convert it to a dataset. Can I put the Batch directly in N in this way to perform the operation. Will this have any impact on the performance of my model. Other than that, do you have any better way to handle this situation?
Here is my code. 
Best Wish.
```
class HyperGCN(nn.Module):
    def __init__(self, in_c, out_c):
        super(HyperGCN, self).__init__()
        self.hgcn = HypergraphConv(in_c, out_c)
        
    def forward(self, x, hyperedge_index, hyperedge_weight):
        # X - [B, N, F],   hyperedge_index- [B, N, M], ,   hyperedge_index- [B, M]
        B, N, F = x.shape
        x = x.reshape(B*N, F)
        y = self.hgcn(x, hyperedge_index , hyperedge_weight)
        return y
```

**Comment by rusty1s** [comment]
This does not work. You also need to take care of increasing the indices of entry in `hyper_edge_index` by the accumulated sum of previous entries.

**Comment by WangYibucea** [comment]
Thank your very much. Could you give us an example of the code so that we can understand it better?

**Comment by rusty1s** [comment]
I'm not entirely sure how your `hyperedge_index` looks like. It seems to be dense while PyG expects it to be sparsely given. In case it is a `[B, 2, E]` matrix containing sparse non-zero indices, you would need to increment `hyperedge_index[:, 0]` by `N` and `hyperedge_index[:, 1]` by `M`:
```
cumsum_N = torch.arange(0, B * N, N)
cumsum_M = torch.arange(0, B * M, M)
cumsum = torch.stack([cumsum_N, cumsum_M], dim=1).view(B, 2, 1)

hyperedge_index = (hyperedge_index + cumsum).permute(1, 0, 2).view(2, -1)
```

**Comment by andac-demir** [comment]
- I also feel the concept of minibatch training for various graph convolutional operators is extremely complicated. 
- Most tutorial scripts use publicly available graph datasets and perform node classification on them. It is difficult to find code examples for graph classification or edge classification on custom datasets. 
- I think only rusty1s maintains this repository, and it is too much responsibility for a single person to fix all these issues.

**Comment by andac-demir** [comment]
import torch
from torch.nn import functional as F
from torch_geometric.data import Data
from torch_geometric.nn import GATConv
from torch_geometric.datasets import Planetoid
import torch_geometric.transforms as T
import matplotlib.pyplot as plt


class GAT(torch.nn.Module):
    def __init__(self, num_node, num_node_feature, hidden, num_class, act, dropout, batch_size):
        super(GAT, self).__init__()
        self.num_node = num_node
        self.num_node_feature = num_node_feature
        self.batch_size = batch_size
        self.act = act
        self.dropout = dropout
        self.in_head = 8
        self.out_head = 1
        self.conv1 = GATConv(self.num_node_feature, hidden, heads=self.in_head, dropout=self.dropout)
        self.conv2 = GATConv(hidden * self.in_head, num_class, concat=False, heads=self.out_head, dropout=self.dropout)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        # x -> 2708 nodes x 1433 features and the dataset has only one graph if dataset is Cora (node classification)a
        # edge_index -> torch.Size([2, 10556])
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv1(x, edge_index)
        # x -> 2708 nodes x (hidden * in_head)
        x = self.act(x)
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv2(x, edge_index)
        # x -> 2708 nodes x 7 classes (7 classes per node if dataset is Cora)
        return F.log_softmax(x, dim=1)


def main():
    name_data = 'Cora'
    dataset = Planetoid(root='/tmp/' + name_data, name=name_data)
    dataset.transform = T.NormalizeFeatures()

    print(f"Number of Classes in {name_data}:", dataset.num_classes)
    print(f"Number of Node Features in {name_data}:", dataset.num_node_features)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    model = GAT(num_node=2708,
                num_node_feature=dataset.num_node_features,
                hidden=8,
                num_class=dataset.num_classes,
                act=torch.nn.ELU(),
                dropout=0.6,
                batch_size=1
                ).to(device)

    data = dataset[0].to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)

    model.train()
    for epoch in range(1000):
        model.train()
        optimizer.zero_grad()
        out = model(data)
        loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])

        if epoch % 200 == 0:
            print(50 * '-')
            print('Epoch: {}, Training Loss: {:.7f}'.format(epoch, loss))
            # Testing
            model.eval()
            _, pred = model(data).max(dim=1)
            correct = float(pred[data.test_mask].eq(data.y[data.test_mask]).sum().item())
            acc = correct / data.test_mask.sum().item()
            print('Test Accuracy: {:.4f}'.format(acc))

        loss.backward()
        optimizer.step()


if __name__ == "__main__":
    main()

**Comment by rusty1s** [comment]
I'm not sure to which issues you are referring, it is entirely intended to be this way :)

Mini-batching is best done using the `DataLoader`, and we have various examples of this, e.g.:
* `mnist_nn_conv.py`: Graph classification on MNIST superpixels
* `point_transformer_classification.py`: Point cloud classification
* `mutag_gin.py`: Graph classification on MUTAG
* `gcn2_ppi.py`: Inductive node classification on multiple graphs
* ...

The issue in this thread seems to resolve around how we can bring a set of graphs into PyG's mini-batch format that already has the same number of nodes and edges (which is extremely rare).

**Comment by andac-demir** [comment]
For instance, I am trying to modify this example which does node classification using GAT Conv in a way to classify graphs in a minibatch training scheme. I can't really make it work. 

```
import torch
from torch.nn import functional as F
from torch_geometric.data import Data
from torch_geometric.nn import GATConv
from torch_geometric.datasets import Planetoid
import torch_geometric.transforms as T
import matplotlib.pyplot as plt


class GAT(torch.nn.Module):
    def __init__(self, num_node, num_node_feature, hidden, num_class, act, dropout, batch_size):
        super(GAT, self).__init__()
        self.num_node = num_node
        self.num_node_feature = num_node_feature
        self.batch_size = batch_size
        self.act = act
        self.dropout = dropout
        self.in_head = 8
        self.out_head = 1
        self.conv1 = GATConv(self.num_node_feature, hidden, heads=self.in_head, dropout=self.dropout)
        self.conv2 = GATConv(hidden * self.in_head, num_class, concat=False, heads=self.out_head, dropout=self.dropout)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        # x -> 2708 nodes x 1433 features and the dataset has only one graph if dataset is Cora (node classification)a
        # edge_index -> torch.Size([2, 10556])
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv1(x, edge_index)
        # x -> 2708 nodes x (hidden * in_head)
        x = self.act(x)
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv2(x, edge_index)
        # x -> 2708 nodes x 7 classes (7 classes per node if dataset is Cora)
        return F.log_softmax(x, dim=1)


def main():
    name_data = 'Cora'
    dataset = Planetoid(root='/tmp/' + name_data, name=name_data)
    dataset.transform = T.NormalizeFeatures()

    print(f"Number of Classes in {name_data}:", dataset.num_classes)
    print(f"Number of Node Features in {name_data}:", dataset.num_node_features)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

    model = GAT(num_node=2708,
                num_node_feature=dataset.num_node_features,
                hidden=8,
                num_class=dataset.num_classes,
                act=torch.nn.ELU(),
                dropout=0.6,
                batch_size=1
                ).to(device)

    data = dataset[0].to(device)
    optimizer = torch.optim.Adam(model.parameters(), lr=0.005, weight_decay=5e-4)

    model.train()
    for epoch in range(1000):
        model.train()
        optimizer.zero_grad()
        out = model(data)
        loss = F.nll_loss(out[data.train_mask], data.y[data.train_mask])

        if epoch % 200 == 0:
            print(50 * '-')
            print('Epoch: {}, Training Loss: {:.7f}'.format(epoch, loss))
            # Testing
            model.eval()
            _, pred = model(data).max(dim=1)
            correct = float(pred[data.test_mask].eq(data.y[data.test_mask]).sum().item())
            acc = correct / data.test_mask.sum().item()
            print('Test Accuracy: {:.4f}'.format(acc))

        loss.backward()
        optimizer.step()



if __name__ == "__main__":
    main()
```

**Comment by andac-demir** [comment]
After modifications (x has to be a torch.Tensor(num_node x num_node_features). I must pass it to the forward method as torch.Tensor(batch_size x num_node x num_node_features)):

```
import torch
from torch import nn
from torch.nn import functional as F
from torch_geometric.data import Data
from torch_geometric.nn import GATConv
from torch_geometric.datasets import Planetoid
import torch_geometric.transforms as T
import matplotlib.pyplot as plt


class GAT(torch.nn.Module):
    def __init__(self, num_node, num_node_feature, hidden, num_class, act, dropout, batch_size):
        super(GAT, self).__init__()
        self.num_node = num_node
        self.num_node_feature = num_node_feature
        self.num_edges = (num_node * (num_node - 1)) // 2  # number of edges in a complete graph
        self.batch_size = batch_size
        self.act = act
        self.dropout = dropout
        self.in_head = 8
        self.out_head = 1
        self.fc_hidden = 100

        # Graph attention convolution
        self.conv1 = GATConv(self.num_node_feature, hidden, heads=self.in_head, dropout=self.dropout)
        self.conv2 = GATConv(hidden * self.in_head, hidden, concat=False, heads=self.out_head, dropout=self.dropout)

        # Fully-connected layers
        self.fc1 = nn.Linear(self.num_node * hidden, self.fc_hidden)  # a dense layer as the read-out function
        self.fc2 = nn.Linear(self.fc_hidden, num_class)

    def forward(self, data):
        x, edge_index, batch = data.x, data.edge_index, data.batch
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv1(x, edge_index)
        x = self.act(x)
        x = F.dropout(x, p=self.dropout, training=self.training)
        x = self.conv2(x, edge_index)
        x = torch.flatten(x)
        # Fully-connected layer
        x = self.act(self.fc1(x))
        emb = x  # graph embedding after read-out function
        return emb, F.log_softmax(self.fc2(x).view(1, -1), dim=-1)
```

**Comment by rusty1s** [comment]
You are missing a global readout in your architecture (instead of flattening), e.g., `global_mean_pool`.

**Comment by andac-demir** [comment]
Hmm, so X is torch.Tensor(batch_size * num_nodes, num_node_features) instead of torch.Tensor(batch_size, num_nodes, num_node_features)? Because GATConv has an assert statement: assert len(x.size()) == 2

**Comment by rusty1s** [comment]
Yes, GATConv does not work in a static graph scenario, and you will need to use regular PyG mini-batching for this, e.g., via Batch.from_data_list or via the DataLoader.

**Comment by andac-demir** [comment]
I see. I have something like this:

```
from torch_geometric.data import Data
from torch_geometric.loader import DataLoader

train_loader = list()
        for i in range(len(train_data)):
            graph = Data(x=train_data[i], y=train_labels[i], edge_index=edge_index)
            train_loader.append(graph)
train_loader = DataLoader(train_loader, batch_size=32, shuffle=True, drop_last=True)
```

train_data is a 3d tensor -> num_samples x num_nodes x num_node_features
I convert all samples to a PyG graph and then minibatch train_loader, which is a list of graphs via DataLoader.

I really appreciate your feedbacks. These are very helpful. Can you please clarify how edge index matrix is modified?

**Comment by WangYibucea** [comment]
I modified my edge, but I think this one is wrong not as a hyperedge, but as a normal edge. The instance as follow：
```
hyperedge_index = torch.ones(B, N, M)
HE = []
for i in range(B):
    edge_index = hyperedge_index[i, :, :].nonzero(as_tuple=False).t()
    HE.append(edge_index.unsqueeze(0))
HE = torch.cat(HE, dim=0)
HEW = []
for i in range(B):
    edge_weight = hyperedge_index[i, HE[i, 0], HE[i, 1]]
    HEW.append(edge_weight.unsqueeze(0))
HEW = torch.cat(HEW, dim=0)
print(HEW.shape)
```

**Comment by rusty1s** [comment]
Yes, the missing part is to increase the indices for each example:
```python
import torch

B = 2
N = 3
M = 4

hyperedge_index = torch.ones(B, N, M)
HE = []
for i in range(B):
    edge_index = hyperedge_index[i, :, :].nonzero(as_tuple=False).t()
    edge_index[0] += i * N
    edge_index[1] += i * M
    HE.append(edge_index)
HE = torch.cat(HE, dim=1)
print(HE.shape)

HEW = []
for i in range(B):
    edge_weight = hyperedge_index[i, :, :].view(-1)
    HEW.append(edge_weight)
HEW = torch.cat(HEW, dim=0)
print(HEW.shape)
```

**Comment by WangYibucea** [comment]
Thank you very much for your reply, I understand the principle inside now. 
But there are still some problems, I have deployed your code to my adaptive hypergraph test code. I found that they report errors in the `Hyperedge_weight` upper dimension, which is most likely caused by the all-1 matrix used before. 
At the moment I am not quite sure what the error is, could you please help me to check？
```
class PAME(nn.Module):
    def __init__(self, DEVICE, t_len, h_feature, o_feature, num_of_vertices):
        super(PAME, self).__init__()
        self.linear1 = nn.Linear(in_features=t_len, out_features=h_feature)
        self.linear2 = nn.Linear(in_features=h_feature, out_features=o_feature)
        self.nodevec = nn.Parameter(torch.randn(num_of_vertices, o_feature), requires_grad=True).to(DEVICE)
        self.edgevec = nn.Parameter(torch.randn(math.ceil(0.3 * num_of_vertices), o_feature), requires_grad=True).to(DEVICE)

    def forward(self, x):
        f = torch.relu(self.linear2(torch.relu(self.linear1(x))))
        DE = torch.tanh(2 * torch.mul(f, self.nodevec))
        EE = torch.tanh(2 * self.edgevec).transpose(1, 0)
        pame = F.relu(torch.tanh(2 * (torch.matmul(DE, EE))))
        return pame


class HyperGCN(nn.Module):
    def __init__(self, in_c, out_c):
        super(HyperGCN, self).__init__()
        self.hgcn = HypergraphConv(in_c, out_c)

    def forward(self, x, hyperedge_index):
        # X - [B, N, F], hyperedge_index- [2, B*E], hyperedge_index- [B*M]
        HE = []
        B, N, M = hyperedge_index.shape
        for i in range(B):
            edge_index = hyperedge_index[i, :, :].nonzero(as_tuple=False).t()
            edge_index[0] += i * N
            edge_index[1] += i * M
            HE.append(edge_index)
        HE = torch.cat(HE, dim=1)
        HEW = []
        for i in range(B):
            edge_weight = hyperedge_index[i, :, :].view(-1)
            HEW.append(edge_weight)
        HEW = torch.cat(HEW, dim=0).nonzero().view(-1)
        B, N, F = x.shape
        x = x.reshape(B*N, F)
        y = self.hgcn(x=x, hyperedge_index=HE, hyperedge_weight=HEW)
        return y
```

**Comment by WangYibucea** [comment]
Thank you very much!!!!!My thesis is saved. Best Wish!!!!

**Comment by 280777510** [comment]
> Hello, thank you very much for your answer. I am currently using HyperGCN. since the graph structure is generated during the training process, it is not convenient to convert it to a dataset. Can I put the Batch directly in N in this way to perform the operation. Will this have any impact on the performance of my model. Other than that, do you have any better way to handle this situation? Here is my code. Best Wish.
> 
> ```
> class HyperGCN(nn.Module):
>     def __init__(self, in_c, out_c):
>         super(HyperGCN, self).__init__()
>         self.hgcn = HypergraphConv(in_c, out_c)
>         
>     def forward(self, x, hyperedge_index, hyperedge_weight):
>         # X - [B, N, F],   hyperedge_index- [B, N, M], ,   hyperedge_index- [B, M]
>         B, N, F = x.shape
>         x = x.reshape(B*N, F)
>         B, N, M = hyperedge_index.shape
>         hyperedge_index = hyperedge_index.reshape(B*N, M)
>         hyperedge_index = hyperedge_index.reshape(B*M)
>         y = self.hgcn(x, hyperedge_index , hyperedge_weight)
>         return y
> ```

did you ever see tutorial?hyperedge_index is not used like this!

**Comment by QQQYang** [comment]
> I see. I have something like this:
> 
> ```
> from torch_geometric.data import Data
> from torch_geometric.loader import DataLoader
> 
> train_loader = list()
>         for i in range(len(train_data)):
>             graph = Data(x=train_data[i], y=train_labels[i], edge_index=edge_index)
>             train_loader.append(graph)
> train_loader = DataLoader(train_loader, batch_size=32, shuffle=True, drop_last=True)
> ```
> 
> train_data is a 3d tensor -> num_samples x num_nodes x num_node_features I convert all samples to a PyG graph and then split train_loader (which is a list of PyG graphs) into minibatches via DataLoader.
> 
> I really appreciate your feedbacks. These are very helpful. Can you please clarify how edge index matrix is modified to train in minibatches? Currently edge_index -> torch.Tensor([2, num_edges]). If I have a minibatch of graphs, is it going to be torch.Tensor([batch_size * 2, num_edges])? I'm lost here.

Hi, I am also confused about this. Have you solved it?

**Comment by mgao97** [comment]
Btw, could pyg support batch sampling on nodes within a global graph for node classification tasks?

**Comment by akihironitta** [comment]
@mgao97 PyG supports neighbour sampling and other sampling methods as documented here https://pytorch-geometric.readthedocs.io/en/2.3.1/modules/loader.html#torch_geometric.loader.NeighborLoader
